gk/fib.c

```diff
@@ -957,14 +957,14 @@ ether_cache_put(struct gk_fib *neigh_fib,
 	int ret, ref_cnt;
 	struct gk_fib *neighbor_fib = neigh_fib;
 
-	ref_cnt = rte_atomic32_sub_return(&eth_cache->ref_cnt, 1);
-	if (ref_cnt < 0) {
+	ref_cnt = rte_atomic32_read(&eth_cache->ref_cnt);
+	if (ref_cnt < 1) {
 		rte_panic("Unexpected condition: the ref_cnt of the ether cache is negative at %s\n",
 			__func__);
-	}
-
-	if (ref_cnt > 0)
+	} else if (ref_cnt >= 2) {
+		rte_atomic32_dec(&eth_cache->ref_cnt);
```

Comment:
We cannot break `rte_atomic32_sub_return()` into `rte_atomic32_read()` and `rte_atomic32_dec()` because the reference counter may have change in between those two calls.

Reply:
Suggestion: use `rte_atomic32_cmpset()`.

Reply:
Could you please explain a little bit more on using ``rte_atomic32_cmpset()``? It seems that we have to subtract 1 from ``ref_cnt`` if it's greater than 1, which indicates that we may have to use two rte_atomic32_xxx functions.

Reply:
We still need to use two rte_atomic32_xxx functions, but `rte_atomic32_cmpset()` let you test that what you read before decrementing is still the case.

Reply:
The final solution will look like this:
```
while ((ref_cnt = rte_atomic32_read(&eth_cache->ref_cnt)) >= 2) {
   if (rte_atomic32_cmpset(&eth_cache->ref_cnt, ref_cnt, ref_cnt - 1) == 0)
      continue;
   return 0;
}
```

Reply:
Polishing up:
```
while ((ref_cnt = rte_atomic32_read(&eth_cache->ref_cnt)) >= 2) {
   if (likely(rte_atomic32_cmpset(&eth_cache->ref_cnt, ref_cnt, ref_cnt - 1) != 0))
      return 0;
}
```